Ds 2486 reindex solr

aschweer opened the pull request.

aschweer committed:
DS-2486 New scripts to export/clear/import solr indexes

aschweer committed:
DS-2486 Add UUID processing to CSV, JSON update handlers too

aschweer committed:
DS-2486 Add Solr import/export to launcher.xml

aschweer commented:
This is an updated PR using a feature branch so that others can add commits if they like. The original PR is here: https://github.com/DSpace/DSpace/pull/894

aschweer commented:
To fix a statistics index from before DSpace 5.x:

* do the maven + ant steps with the DSpace 5.x codebase
* start Tomcat
* run `[dspace]/bin/dspace solr-reindex-statistics`
* add the `--keep` flag if you wish to keep the exported CSV files; by default, they will be deleted after the reindex is complete

To export statistics data for back-up purposes (substitute "authority" for "statistics" to export the Solr-based authority data):

* with Tomcat running, run `[dspace]/bin/dspace solr-export-statistics`
* optionally, use `--last d` to export just 1 day's worth of statistics or `--last 7` for 7 days' worth of statistics (can use any non-negative number of days) or `--last m` for previous month's statistics
* it will export to `[dspace]/solr-export` unless an alternative directory is specified with the `--directory` option

To import previously backed-up statistics (substitute "authority" for "statistics" to export the Solr-based authority data):

* with Tomcat running, run `[dspace]/bin/dspace solr-import-statistics`
* it will import from `[dspace]/solr-export` unless an alternative directory is specified with the `--directory` option (this should match what you used for the export)
* it will add the imported data to the existing index unless the `--clear` flag is passed in

The export will take up approximately as much disk space as the index's data directory. The reindex temporarily swaps out cores and needs some space in `[dspace]/temp` to store statistics hits / authority information coming in while the export portion of the reindex is happening.

hardyoyo commented:
I am attempting to run this on my staging server, with a copy of my live server's statistics core in place. I can confirm the export seems to be working fine, and the site continues to function normally. That is pretty exciting. I will let you know when it is done re-indexing, and whether the geo data returns.

hardyoyo commented:
ran into trouble at some point, looks like the Solr server stopped for some reason during the process (it might be that a nightly crontab interrupted the process... I will re-try with a fresh copy of the stats core. I will paste the errors here, just to keep a record of them:

bin/dspace solr-reindex-statistics -a reindex -d /var2/solr-export
java.io.IOException: Server returned HTTP response code: 500 for URL: http://localhost:8080/solr/statistics-temp/select?q=*%3A*&rows=10000&wt=csv&fl=*&sort=time+asc&start=7670000
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1627)
        at java.net.URL.openStream(URL.java:1037)
        at org.apache.commons.io.FileUtils.copyURLToFile(FileUtils.java:1392)
        at org.dspace.util.SolrImportExport.exportIndex(SolrImportExport.java:456)
        at org.dspace.util.SolrImportExport.exportIndex(SolrImportExport.java:282)
        at org.dspace.util.SolrImportExport.reindex(SolrImportExport.java:230)
        at org.dspace.util.SolrImportExport.main(SolrImportExport.java:131)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.dspace.app.launcher.ScriptLauncher.runOneCommand(ScriptLauncher.java:225)
        at org.dspace.app.launcher.ScriptLauncher.main(ScriptLauncher.java:77)
[dspace@lso-dev-01 dspace]$ 
[dspace@lso-dev-01 dspace]$ 
[dspace@lso-dev-01 dspace]$ 
[dspace@lso-dev-01 dspace]$ ^C
[dspace@lso-dev-01 dspace]$ pwd
/dspace
[dspace@lso-dev-01 dspace]$ ^C
[dspace@lso-dev-01 dspace]$ 
[dspace@lso-dev-01 dspace]$ pwd
/dspace
[dspace@lso-dev-01 dspace]$ bin/dspace solr-reindex-statistics -a reindex -d /var2/solr-export                       
Exception: Core with name 'statistics-temp' already exists.
org.apache.solr.client.solrj.impl.HttpSolrServer$RemoteSolrException: Core with name 'statistics-temp' already exists.
        at org.apache.solr.client.solrj.impl.HttpSolrServer.executeMethod(HttpSolrServer.java:552)
        at org.apache.solr.client.solrj.impl.HttpSolrServer.request(HttpSolrServer.java:210)
        at org.apache.solr.client.solrj.impl.HttpSolrServer.request(HttpSolrServer.java:206)
        at org.apache.solr.client.solrj.request.CoreAdminRequest.process(CoreAdminRequest.java:510)
        at org.dspace.util.SolrImportExport.reindex(SolrImportExport.java:215)
        at org.dspace.util.SolrImportExport.main(SolrImportExport.java:131)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.dspace.app.launcher.ScriptLauncher.runOneCommand(ScriptLauncher.java:225)
        at org.dspace.app.launcher.ScriptLauncher.main(ScriptLauncher.java:77)

hardyoyo commented:
Hmm... not sure what's going on here, I will see if I can find something in catalina.out. The reindex ran most of the day, but stopped just now with these errors: java.io.IOException: Server returned HTTP response code: 500 for URL: http://localhost:8080/solr/statistics-temp/select?q=*%3A*&rows=10000&wt=csv&fl=*&sort=time+asc&start=7700000
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1627)
        at java.net.URL.openStream(URL.java:1037)
        at org.apache.commons.io.FileUtils.copyURLToFile(FileUtils.java:1392)
        at org.dspace.util.SolrImportExport.exportIndex(SolrImportExport.java:456)
        at org.dspace.util.SolrImportExport.exportIndex(SolrImportExport.java:282)
        at org.dspace.util.SolrImportExport.reindex(SolrImportExport.java:230)
        at org.dspace.util.SolrImportExport.main(SolrImportExport.java:131)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.dspace.app.launcher.ScriptLauncher.runOneCommand(ScriptLauncher.java:225)
        at org.dspace.app.launcher.ScriptLauncher.main(ScriptLauncher.java:77)

hardyoyo commented:
OK, it looks like Tomcat ran out of heap space at some point:

java.lang.OutOfMemoryError: Java heap space
Dumping heap to java_pid46960.hprof ...
Unable to create java_pid46960.hprof: Permission denied
Exception in thread "http-bio-127.0.0.1-8080-exec-11" java.lang.OutOfMemoryError: Java heap space
        at org.apache.solr.search.SolrIndexSearcher.getDocListNC(SolrIndexSearcher.java:1627)
        at org.apache.solr.search.SolrIndexSearcher.getDocListC(SolrIndexSearcher.java:1433)
        at org.apache.solr.search.SolrIndexSearcher.search(SolrIndexSearcher.java:514)
        at org.apache.solr.handler.component.QueryComponent.process(QueryComponent.java:485)
        at org.apache.solr.handler.component.SearchHandler.handleRequestBody(SearchHandler.java:218)
        at org.apache.solr.handler.RequestHandlerBase.handleRequest(RequestHandlerBase.java:135)
        at org.apache.solr.core.SolrCore.execute(SolrCore.java:1967)
        at org.apache.solr.servlet.SolrDispatchFilter.execute(SolrDispatchFilter.java:777)
        at org.apache.solr.servlet.SolrDispatchFilter.doFilter(SolrDispatchFilter.java:418)
        at org.apache.solr.servlet.SolrDispatchFilter.doFilter(SolrDispatchFilter.java:207)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
        at org.dspace.solr.filters.LocalHostRestrictionFilter.doFilter(LocalHostRestrictionFilter.java:50)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:220)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:122)
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:503)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:170)
        at com.googlecode.psiprobe.Tomcat70AgentValve.invoke(Tomcat70AgentValve.java:38)
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:103)
        at org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:950)
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:116)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:421)
        at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1070)
        at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:611)
        at org.apache.tomcat.util.net.JIoEndpoint$SocketProcessor.run(JIoEndpoint.java:316)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
        at java.lang.Thread.run(Thread.java:745)

Perhaps I need to beef up the memory allocation for my tomcat server so it can survive this reindex attempt. Here is what I have right now:

export CATALINA_OPTS="$CATALINA_OPTS -Xmx2560M"
export CATALINA_OPTS="$CATALINA_OPTS -Xms1536M"
export CATALINA_OPTS="$CATALINA_OPTS -XX:MaxPermSize=1024M"
export CATALINA_OPTS="$CATALINA_OPTS -XX:PermSize=820M"
export CATALINA_OPTS="$CATALINA_OPTS -Xss228k"

This is my staging server, it runs several versions of DSpace... I can also try dropping some of the contexts to make room. If anyone has advice for properly tuning for a reindex, post your hints here, thanks!

hardyoyo commented on dspace-api/src/main/java/org/dspace/util/SolrImportExport.java:
```diff
@@ -0,0 +1,583 @@
+	private static final String KEEP_OPTION = "k";
+	private static final String LAST_OPTION = "l";
+
+	public static final int ROWS_PER_FILE = 10_000;
```
I realize this is set to 10K to make the process go faster, but it's fairly demanding of a server that may be doing other things at the same time. I am going to try setting this number lower and see how that works out.

hardyoyo commented:
Well... I left a reindex process running on Friday afternoon, started about 5pm. Confirmed the site was still working, and watched CSV file being created. All looked good. I left that running overnight, and on Saturday, about 6pm (25 hours later) ... unfortunatetly, our sysadmins rebooted the server after loading some patches. From the looks of things, the reindex did not complete (still no geo information in the stats core). I will have to retry my experiment. I would *love* to have some help from anyone else, trying to do the same kind of testing. Surely there is someone else out there getting ready for an upgrade to 5.x?

aschweer commented on dspace-api/src/main/java/org/dspace/util/SolrImportExport.java:
```diff
@@ -0,0 +1,583 @@
+		query.setRows(ROWS_PER_FILE);
+		query.set("wt", "csv");
+		query.set("fl", "*");
+		query.setSort(timeField, SolrQuery.ORDER.asc);
```
The sorting appears to cause the memory issues that people have reported running into when doing the export. I ran the Eclipse Memory Analyzer on a heap dump that @hardyoyo provided, and both problem suspects reported by that tool point to something related to sorting Solr documents. I'm going to try and change the code to process documents in weekly or monthly batches (using filter queries) so that it only has to sort the documents in a given batch.

aschweer committed:
Export monthly batches to limit the number of docs to sort

aschweer committed:
Use stats component to get minimum date

aschweer commented:
Here is a version that processes the data in monthly batches, so that only one month's worth of docs need to be sorted at a time. I haven't tested this beyond making sure that the export actually works. 

I'm wondering whether we can get away with not sorting the documents in one batch at all; there will be offset queries if the month has >10k docs but I don't know whether giving an offset means it's best practice to also sort. Looking at the solr documentation, perhaps https://wiki.apache.org/solr/CommonQueryParameters#Deep_paging_with_cursorMark would be a starting point if sorting the monthly batches still requires a lot of resources.

mwoodiupui commented:
I just tried this (with a much smaller set of event records) and it did indeed restore the geographic statistics.  Original data were from a v1.8.1 instance.

+1

peterdietz commented:
Recreating the problem, I cloned a 4.x site, and imported its DB + /dspace/solr and started tomcat. Here is a good statistics page:
![screen shot 2015-04-16 at 4 26 54 pm](https://cloud.githubusercontent.com/assets/58014/7190621/72f8ae7a-e455-11e4-91f5-f7081499559b.png)

Okay, then to reproduce the problem. I git checkout dspace-5_x, which is essentially 5.1, and run mvn+ant, which upgrades the solr, and flyway updates the DB. Viewing that same collection statistics has the problem. No geo information.
![screen shot 2015-04-16 at 4 28 57 pm](https://cloud.githubusercontent.com/assets/58014/7190656/b55f8a86-e455-11e4-8ab2-421897c7bfba.png)

I think checkout this branch. And run the solr-reindex-statistics
2015-04-16 16:21:59,002 INFO  org.dspace.util.SolrImportExport @ Exported batch 0 to /dspace/solr-export/statistics_export_2011-10_0.csv
2015-04-16 16:21:59,022 INFO  org.dspace.util.SolrImportExport @ Exported batch 10000 to /dspace/solr-export/statistics_export_2011-10_1.csv
2015-04-16 16:21:59,091 INFO  org.dspace.util.SolrImportExport @ Exported batch 0 to /dspace/solr-export/statistics_export_2011-11.csv
2015-04-16 16:21:59,252 INFO  org.dspace.util.SolrImportExport @ Exported batch 0 to /dspace/solr-export/statistics_export_2011-12_0.csv
2015-04-16 16:21:59,418 INFO  org.dspace.util.SolrImportExport @ Exported batch 10000 to /dspace/solr-export/statistics_export_2011-12_1.csv
2015-04-16 16:21:59,478 INFO  org.dspace.util.SolrImportExport @ Exported batch 20000 to /dspace/solr-export/statistics_export_2011-12_2.csv
...
2015-04-16 16:25:47,415 INFO  org.dspace.util.SolrImportExport @ Importing file /dspace/solr-export/statistics_export_2015-03.csv
2015-04-16 16:25:47,435 INFO  org.dspace.util.SolrImportExport @ Importing file /dspace/solr-export/statistics_export_2015-04_0.csv
2015-04-16 16:25:48,135 INFO  org.dspace.util.SolrImportExport @ Importing file /dspace/solr-export/statistics_export_2015-04_1.csv
2015-04-16 16:25:48,867 INFO  org.dspace.util.SolrImportExport @ Importing file /dspace/solr-export/statistics_export_2015-04_2.csv
2015-04-16 16:25:49,523 INFO  org.dspace.util.SolrImportExport @ Importing file /dspace/solr-export/statistics_export_2015-04_3.csv

It got very close to the end... But I just got this exception...

/dspace/bin/dspace solr-reindex-statistics
Exception: null
java.lang.NullPointerException
	at org.dspace.util.SolrImportExport.exportIndex(SolrImportExport.java:446)
	at org.dspace.util.SolrImportExport.exportIndex(SolrImportExport.java:286)
	at org.dspace.util.SolrImportExport.reindex(SolrImportExport.java:251)
	at org.dspace.util.SolrImportExport.main(SolrImportExport.java:135)

The statistics page looks good though, as though the import mostly succeeded. I'll dig into the error in a bit.
![screen shot 2015-04-16 at 4 57 23 pm](https://cloud.githubusercontent.com/assets/58014/7191210/d1d5b68c-e459-11e4-9f1e-749afed03cfa.png)

aschweer commented:
Thanks a lot @mwoodiupui and @peterdietz for testing! How many stats hits you have in your indexes? Something like this: `curl --globoff 'http://localhost:8080/solr/statistics/select?q=*:*&rows=0&indent=true' | grep numFound` Also, how much memory did you give your tomcat when you ran this?

@mwoodiupui, the later commits don't change the basic functionality, but I guess it would still be good to test the actual code that might make it into 5.2 :) @peterdietz, looking at that line, I think this may happen when the monthly batches "overshoot" the current date in some way. I'll look into that.

peterdietz commented:
I would say this is a small/medium instance. The number of Solr stats records was 2,651,202.

hardyoyo commented:
Testing on a recent copy of my production stats core, 17M item views. The export phase is very well behaved, never ran out of memory, and proceeded very quickly. However, the process ended much the same as Peter says, with an NPE:

java.lang.NullPointerException
        at org.dspace.util.SolrImportExport.exportIndex(SolrImportExport.java:446)
        at org.dspace.util.SolrImportExport.exportIndex(SolrImportExport.java:286)
        at org.dspace.util.SolrImportExport.reindex(SolrImportExport.java:251)
        at org.dspace.util.SolrImportExport.main(SolrImportExport.java:135)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.dspace.app.launcher.ScriptLauncher.runOneCommand(ScriptLauncher.java:225)
        at org.dspace.app.launcher.ScriptLauncher.main(ScriptLauncher.java:77)

hardyoyo commented:
I was following the dspace.log file, here are the last few lines from it before the NPE:

2015-04-16 16:07:06,880 INFO  org.dspace.util.SolrImportExport @ Importing file /var2/solr-export/statistics_export_2011-05_13.csv
2015-04-16 16:07:12,630 INFO  org.dspace.util.SolrImportExport @ Importing file /var2/solr-export/statistics_export_2012-03_11.csv
2015-04-16 16:07:16,278 INFO  org.dspace.util.SolrImportExport @ Importing file /var2/solr-export/statistics_export_2013-10_00.csv
2015-04-16 16:07:19,988 INFO  org.dspace.util.SolrImportExport @ Importing file /var2/solr-export/statistics_export_2013-06_015.csv
2015-04-16 16:07:22,534 INFO  org.dspace.util.SolrImportExport @ Importing file /var2/solr-export/statistics_export_2011-05_01.csv
2015-04-16 16:07:25,773 INFO  org.dspace.util.SolrImportExport @ Importing file /var2/solr-export/statistics_export_2014-09_01.csv

hardyoyo commented:
From looking at the logs, the export phase took about 47 minutes to complete, the import phase proceeded about 2 hours before hitting an NPE.

aschweer commented:
Thanks for the extra information and sorry about that NPE! I'll see what I can do.

peterdietz commented:
I got stuck in a weird state. I reset my /dspace/solr directory, and re-ran the script solr-reindex-statistics, but it crashed because the files already existed:
org.dspace.util.SolrImportExportException: Could not create file /dspace/solr-export/statistics_export_2010-12.csv while exporting index statistics, month2010-12-01T00:00:00Z, batch 0

So.. I deleted all the files in that /dspace/solr-export directory, and ran it again.
Exception: Core with name 'statistics-temp' already exists.
org.apache.solr.client.solrj.impl.HttpSolrServer$RemoteSolrException: Core with name 'statistics-temp' already exists.

So. I had to re-reset my /dspace/solr directory, and delete the files in that solr-export directory. Its just a weird state to get stuck in. And I hope that end-users never get stuck there.

aschweer commented:
So the way this works is: 
1. create a temporary core to hold stats hits coming in while the reindex is running
2. swap cores so that the solr logger points to the temporary one
3. export the actual core
4. clear the actual core & import all data from the export
5. swap cores back
6. export the temporary core
7. import the temporary core into the actual core
8. drop the temporary core

It looks like the NPE made you get stuck somewhere in step 6 or 7, so the temporary core would still be sticking around (like you saw, @peterdietz) and messing things up. I think the problem may happen when no actual hits come in between steps 2 and 5, ie when the temporary core is empty. 

I think I also just spotted a problem with step 7 potentially importing too much, so I'll look into that.

hardyoyo commented:
That makes sense, I didn't test whether the site was functioning during the reindex this time, so I didn't make any new item views for the temp core.

aschweer committed:
Make import/export of temporary core more robust

peterdietz commented:
I changed SolrImportExport.java line 446 to have a check:
        Date earliestTimestamp;
        if(fieldInfo.containsKey(timeField)) {
            earliestTimestamp = (Date) fieldInfo.get(timeField).getMin();
        } else {
            log.info("Timefield: " + timeField + " fromWhen: " + fromWhen);
            earliestTimestamp = null;
        }

The fieldInfo.containsKey(timeField) was only false once, and then it crashed.
2015-04-16 17:46:03,539 INFO  org.dspace.util.SolrImportExport @ Importing file /dspace/solr-export/statistics_export_2015-04_3.csv
2015-04-16 17:46:04,177 INFO  org.dspace.util.SolrImportExport @ Timefield: time fromWhen: null

I set earliestTimestamp to null in the else, which will break SimpleDateFormat a bit later.

aschweer commented:
I think when there is no earliest date, the export just needs to stop trying. I just pushed a few tweaks, would you mind giving it another go?

peterdietz commented:
:+1: The reindex (export+import) completed successfully. No errors. Again, this on a smallish 2M records instance.

```
2015-04-16 18:17:07,200 INFO  org.dspace.util.SolrImportExport @ Importing file /dspace/solr-export/statistics_export_2015-04_3.csv
2015-04-16 18:17:07,865 WARN  org.dspace.util.SolrImportExport @ Cannot get earliest date, not exporting index statistics-temp, time field time, from null
2015-04-16 18:17:07,916 WARN  org.dspace.util.SolrImportExport @ No export files found in directory /dspace/temp/solr-data-tmp for index statistics-temp
```

aschweer commented:
Thanks a lot @peterdietz, looks like we're getting there!

aschweer commented:
I run ran the reindex on the stats core of one of "my" repositories that has been upgraded since 1.6 or so and that has about 2.6M statistics hits going back to 2010. The reindex ran through without any problems and took about 25 minutes. This was in a tomcat instance that only has 512MB heap space (but had no traffic during the reindex). So to me it looks like my most recent code changes resolved the performance problems :)

hardyoyo commented:
Indeed, I started another run on a copy of my production stats core, 17M item views written out to CSV in about 50 minutes, import proceeding now. I'll post an update when I check on the progress in the morning.

aschweer commented:
@hardyoyo thanks for the update -- good to hear that the export went through fine, given that this is where the OOM problems came up earlier!

aschweer commented:
Something still isn't right. I just ran this on another stats core with about 7M documents. The export appears to have gone through fine (in under 30 minutes), but I ran into a version conflict when importing. This shouldn't happen; the `_version_` field should be skipped during the import. So I'll have to check what is going on there.

````
Exception: version conflict for 5f165078-ef8b-48a7-ac72-c941bfc45afc expected=1498677320117911552 actual=-1
org.apache.solr.client.solrj.impl.HttpSolrServer$RemoteSolrException: version conflict for 5f165078-ef8b-48a7-ac72-c941bfc45afc expected=1498677320117911552 actual=-1
	at org.apache.solr.client.solrj.impl.HttpSolrServer.executeMethod(HttpSolrServer.java:552)
	at org.apache.solr.client.solrj.impl.HttpSolrServer.request(HttpSolrServer.java:210)
	at org.apache.solr.client.solrj.impl.HttpSolrServer.request(HttpSolrServer.java:206)
	at org.dspace.util.SolrImportExport.importIndex(SolrImportExport.java:356)
	at org.dspace.util.SolrImportExport.reindex(SolrImportExport.java:254)
	at org.dspace.util.SolrImportExport.main(SolrImportExport.java:135)
````
Ah right, the `_version_` field is skipped by my code only when the index is being cleared, which isn't the case when additional stats hits are imported from the temporary core (step 7 in my list above). I think it would be fine to just always skip the `_version_` field -- it would be fine for the reindex, but I may be forgetting a use case where that isn't true.

hardyoyo commented:
I'm seeing the same error, right at the end:

Exception: version conflict for f4687520-7c8e-4d67-9b1e-6c85b1d4e41b expected=1498667267663069184 actual=-1
org.apache.solr.client.solrj.impl.HttpSolrServer$RemoteSolrException: version conflict for f4687520-7c8e-4d67-9b1e-6c85b1d4e41b expected=1498667267663069184 actual=-1
        at org.apache.solr.client.solrj.impl.HttpSolrServer.executeMethod(HttpSolrServer.java:552)
        at org.apache.solr.client.solrj.impl.HttpSolrServer.request(HttpSolrServer.java:210)
        at org.apache.solr.client.solrj.impl.HttpSolrServer.request(HttpSolrServer.java:206)
        at org.dspace.util.SolrImportExport.importIndex(SolrImportExport.java:356)
        at org.dspace.util.SolrImportExport.reindex(SolrImportExport.java:254)
        at org.dspace.util.SolrImportExport.main(SolrImportExport.java:135)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.dspace.app.launcher.ScriptLauncher.runOneCommand(ScriptLauncher.java:225)
        at org.dspace.app.launcher.ScriptLauncher.main(ScriptLauncher.java:77)

I will be on a field trip with my daughter most of the day, but will be able to run another test this evening.

hardyoyo commented:
It looks like you have some code in place for skipping the _version_ field already: https://github.com/DSpace/DSpace/blob/DS-2486-Reindex-Solr/dspace-api/src/main/java/org/dspace/util/SolrImportExport.java#L346 would we perhaps be able to leverage this by adding a skipVersion boolean param to the importIndex metthod?

aschweer committed:
Disable version check when importing temporary stats data

aschweer commented:
@hardyoyo yes definitely -- here's an updated version of the code.

hardyoyo commented:
re-running a test of the latest version (my attempt yesterday failed because I made a dumb mistake, and did a binary/fast build with Maven, by running mvn in dspace-src/dspace, which would of course leave out this change). the export phase just completed in about 45 minutes, I see this line: at the end of the export phase:

2015-04-21 12:48:30,555 WARN  org.dspace.util.SolrImportExport @ Cannot get earliest date, not exporting index statistics-temp, time field time, from null

And now I see importing has begun.

More later, whenever this test completes.

hardyoyo commented:
It looks like the script is done, however, it ended with this error:

java.io.IOException: Unable to delete directory /var2/solr-export.
        at org.apache.commons.io.FileUtils.deleteDirectory(FileUtils.java:1473)
        at org.dspace.util.SolrImportExport.reindex(SolrImportExport.java:267)
        at org.dspace.util.SolrImportExport.main(SolrImportExport.java:135)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.dspace.app.launcher.ScriptLauncher.runOneCommand(ScriptLauncher.java:225)
        at org.dspace.app.launcher.ScriptLauncher.main(ScriptLauncher.java:77)

I think it might be safe to assume that if the directory was provided on the command line that we don't want to delete it. :-) I will see if the stats look re-indexed.

hardyoyo commented:
The reindex script seems to have worked (inspected dspace.log files to confirm, export and import phases completed without error). The cleanup behavior was a bit of a surprise--I added a note to DS-2486 about documenting the use of the --keep option if the user cares about the export folder or its contents. With documentation, this script is ready to go as-is. +1 for merge

mwoodiupui commented on dspace-api/src/main/java/org/dspace/util/SolrImportExport.java:
```diff
@@ -0,0 +1,633 @@
+
+			if (!keepExport)
+			{
+				FileUtils.deleteDirectory(new File(exportDir));
```
Isn't there an ambiguity here?  exportIndex() might or might not have created the directory at exportDir, and we don't know which.  May I suggest that main() should, if necessary, create this directory and remember that it did so, present reindex()/exportIndex()/importIndex() with a guaranteed pre-existing directory, and clean up any directories it created after the index operation is complete.  main() takes care of setup and cleanup, and no other method has to know what has been done elsewhere.

aschweer replied:
Thanks @mwoodiupui  / @hardyoyo , yes that would definitely be cleaner. I'm not sure I'll have time today to change the code, so feel free to push to the PR branch if you have time.

aschweer commented:
I'm just running the reindex on one of my test servers; based on that experience, I think it would be good to catch the most likely problem: the export directory running out of space. The main issue is that the script leaves behind the statistics-temp core. This is just a temporary core, so restarting tomcat gets rid of it (while losing any stats hits that came in since the reindex started), but this isn't obvious from the error message. 

We could query free space (http://docs.oracle.com/javase/7/docs/api/java/nio/file/FileStore.html#getUsableSpace%28%29) before doing anything and warn if it looks like there isn't enough left (eg less than 1.2 times the size of [dspace]/solr/statistics/data). Any thoughts?

vidiecan commented on dspace-api/src/main/java/org/dspace/util/SolrImportExport.java:
```diff
@@ -0,0 +1,633 @@
+		// Create a temp directory to store temporary core data
+		File tempDataDir = new File(ConfigurationManager.getProperty("dspace.dir") + File.separator + "temp" + File.separator + "solr-data");
+		boolean tempDirCreated = tempDataDir.mkdirs();
+		if (!tempDirCreated)
```
Note:
Running ```./dspace solr-reindex-statistics``` as e.g., root, creates temp/ directory with too strict permissions and tomcat will not be able to write into it.
If you create the temp/solr-data directory beforehand, it will throw exception at this line.

aschweer replied:
Honestly, that is not a problem as far as I'm concerned. Running any ./dspace scripts as a user different from the tomcat user can cause problems. You should never do that.

hardyoyo commented:
OK, this is a digression, but I'm wondering if at some point we can add isBot processing to the reindex? That would be a handy way to address DS-2431 (https://jira.duraspace.org/browse/DS-2431).

aschweer committed:
[DS-2486] Increase robustness, improve directory delete behaviour

As suggested by Mark Wood, delete directories only when they were actually
created (and when --keep is not given).

Also refuse to go ahead with reindexing when it's obvious that there won't be
enough space, plus deal a little more gracefully with common errors (initial
export failing; -temp core still left from previous attempt).

aschweer commented:
I've changed the behaviour somewhat around deleting directories. main() now ensures that the directory exists and is readable/writable for import/export. For reindex, directories are deleted only if they were created by the method. There is one thing left that I'm not sure about; when the directory exists during reindex, the exported files are kept around even when --keep is _not_ given. The alternative would be to delete those files that look like they have been created by the reindex; I'm not sure that's particularly elegant.

Another change is that the reindex code now tries to determine the Solr core size (via the core admin 'status' command) and refuses to go ahead if the export directory doesn't have at least that much usable space left. I'm not sure how reliable this is and it may cause more problems than it prevents. However, running out of space in the export directory feels like the most probable reason why the reindex may fail (after permissions issues, and those are checked first), so I hope this behaviour is reasonable/helpful.

I haven't done a lot of testing of these changes, so if anyone has a bit of time to give these a go, that'd be great.

aschweer commented:
Also, just a comment regarding the `--keep` flag -- to me, the default should be to delete the exported files when reindexing. I see the reindexing use case as "I don't care what happens, just make my stats work"; in that case, having potentially several GB of "useless" files sticking around would surprise me. When importing/exporting, the `--keep` flag isn't used anyway. 

Obviously others can disagree with what is and isn't surprising :) So if the majority here thinks that reindexing should keep the exported files then please say so.

hardyoyo commented:
I like the new additions, will test them later today. A question: would you recommend running stats-util -o after running a reindex operation and/or an import? If so, we should add it to the docs.

aschweer commented:
Thanks @hardyoyo. Good question regarding the optimise. To be honest, I don't know enough about the effect of optimising a solr core to have an informed opinion on this. It does look like optimise isn't always a good thing: http://blog.trifork.com/2011/11/21/simon-says-optimize-is-bad-for-you/ and also https://wiki.apache.org/solr/SolrPerformanceFactors#Optimization_Considerations

hardyoyo commented:
hmm.. those two links make my head hurt... :-) it sounds like we don't get much from an optimization, though I think we still recommend running stats-util -o every night via crontab: https://wiki.duraspace.org/display/DSDOC5x/Scheduled+Tasks+via+Cron

aschweer commented:
Just from the Solr work I have been doing, it looks to me like the main benefit in optimising is after deleting some documents. My code already optimizes the index after clearing (https://github.com/DSpace/DSpace/pull/905/files#diff-8598adb36e720567e6544bb48f6ce0b9R492). Which suggests that optimising isn't needed again after the reindex (nor nightly really, for that matter, but we may wish to throw it in automatically after stats-util -i and the like).

hardyoyo commented:
OK, I'm testing this script with my "newb" hat on, and I've made a simple 100MB loop/temp file mount for testing, and the first test failed because of permissions:

[dspace@lso-dev-01 dspace]$ bin/dspace solr-reindex-statistics -a reindex -d /mnt/test_loop
org.dspace.util.SolrImportExportException: Can't write to export directory /mnt/test_loop
        at org.dspace.util.SolrImportExport.reindex(SolrImportExport.java:243)
        at org.dspace.util.SolrImportExport.main(SolrImportExport.java:165)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.dspace.app.launcher.ScriptLauncher.runOneCommand(ScriptLauncher.java:225)
        at org.dspace.app.launcher.ScriptLauncher.main(ScriptLauncher.java:77)

I'll give proper permissions to this folder before further testing, but I thought you might find this failure interesting. More later.

hardyoyo commented:
It looks like the check for export filesystem size isn't working:

[dspace@lso-dev-01 dspace]$ bin/dspace solr-reindex-statistics -a reindex -d /mnt/test_loop
Encountered problem during reindex: No space left on device, will attempt to restore Solr cores
java.io.IOException: No space left on device
        at java.io.FileOutputStream.writeBytes(Native Method)
        at java.io.FileOutputStream.write(FileOutputStream.java:345)
        at org.apache.commons.io.IOUtils.copyLarge(IOUtils.java:1720)
        at org.apache.commons.io.IOUtils.copyLarge(IOUtils.java:1696)
        at org.apache.commons.io.IOUtils.copy(IOUtils.java:1671)
        at org.apache.commons.io.FileUtils.copyInputStreamToFile(FileUtils.java:1444)
        at org.apache.commons.io.FileUtils.copyURLToFile(FileUtils.java:1393)
        at org.dspace.util.SolrImportExport.exportIndex(SolrImportExport.java:589)
        at org.dspace.util.SolrImportExport.exportIndex(SolrImportExport.java:371)
        at org.dspace.util.SolrImportExport.reindex(SolrImportExport.java:306)
        at org.dspace.util.SolrImportExport.main(SolrImportExport.java:165)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.dspace.app.launcher.ScriptLauncher.runOneCommand(ScriptLauncher.java:225)
        at org.dspace.app.launcher.ScriptLauncher.main(ScriptLauncher.java:77)

hardyoyo commented:
The actual reindex seems to work fine, when pointed to a folder that has enough slack space and the proper permissions. I'ved reduced the size of my stats core from 17M to 11M (dropped 6M hits from a Google Search Appliance), so this run is going much faster than before.

hardyoyo commented:
Reindex completed successfully, geographic data is present in the stats index. I specified a folder using the -d option, the script did not attempt to delete it this time. The exported csv files remained, which is fine, we might tell people that they will remain after a reindex. I'm pleased with the results, however, you might want to tinker with the safeguard checks you put into place. I can re-test whenever you wish.

aschweer commented:
@hardyoyo thanks for testing. I don't think it's possible to get the disk space right 100%; I don't think it's possible to predict the exact size of the CSV export from the size of the binary Solr index. At the moment, the code is complaining if the export directory doesn't have at least as much space (at the time the export is started) as the size of the index. We can throw in a factor 1.5 or 2 or whatever seems reasonable, but at the end of the day that will always just be a guess. And if the export takes a long time then the disk may fill up with other stuff in the meantime.

To get some more data, could you check what the total size of your CSV export is and also let me know what the size on disk is of the solr/statistics/data directory?

aschweer commented:
@hardyoyo when it actually ran out of disk space, did it sort out the solr cores? As in, could you run it again after freeing up space / pointing it elsewhere _without_ getting warnings about the statistics-temp core?

hardyoyo commented:
@aschweer Yes, it ran fine after running out of disk space. BTW, my loopback test mount was formatted to 100MB, significantly smaller than it needed to be. Mark Wood and I tried to puzzle it out earlier today, we're not entirely sure about the method of getting the size in bytes via the core admin. I think we may have to hook up a debugger to figure out what's happening, I haven't had a chance to do so, yet.

aschweer commented:
@hardyoyo good to hear that it recovered from the problem at least.

You can get the size via curl as well, if you want to compare binary size to CSV export / check whether the code is doing the right thing: 

    curl --globoff 'http://localhost:8080/solr/admin/cores?core=statistics&indent=true'

hardyoyo commented:
Mark and I were looking at this line

https://github.com/DSpace/DSpace/pull/905/files#diff-8598adb36e720567e6544bb48f6ce0b9R255

and wondering why it might have failed, since usableExportSpace was clearly not big enough to hold the exports (in my test), and we figured that something must be a bit off farther upstream, and we had trouble tracing how https://github.com/DSpace/DSpace/pull/905/files#diff-8598adb36e720567e6544bb48f6ce0b9R251 and https://github.com/DSpace/DSpace/pull/905/files#diff-8598adb36e720567e6544bb48f6ce0b9R252 were working (the Solr docs around this area aren't great, we were looking at the actual Solr code to see if we could figure it out, and ended up deciding it was a job for a debugger.

aschweer commented:
I believe the curl query above is the equivalent of those Java calls, so you could use that for a quick sanity check if you don't have time for a debugger.

hardyoyo commented:
Even without working safeguards, I'm +1 this script, it gets the job done, you just have to keep your wits about you. I'm OK with merging it as-is and fixing it later.

peterdietz commented:
+1, the script does what it needs to do. Thanks for the fix.

The user should keep in mind that they should ensure they have a backup of their solr statistics, just in case.

aschweer commented:
I definitely agree that we should recommend that people make a backup of their solr stats data dir before running this.

That's +1s from two other committers and I think the code is as stable as it will get for now. I'll push the scary green button.

aschweer commented:
Oh, and I forgot to say: thanks a lot to all of you who worked on this with me!

peterdietz commented:
This one went into master, 6.x+ Do we need to create another PR to address 5.x?

hardyoyo commented:
I just backported it to the 5_x branch, as is our custom.
